core/storage/btree.rs

```diff
@@ -6604,8 +6604,114 @@ fn free_cell_range(
     Ok(())
 }
 
+/// This function handles pages with two or fewer freeblocks and max_frag_bytes (parameter to defragment_page())
+/// or fewer fragmented bytes. In this case it is faster to move the two (or one)
+/// blocks of cells using memmove() and add the required offsets to each pointer
+/// in the cell-pointer array than it is to reconstruct the entire page.
+/// Note that this function will leave max_frag_bytes as is, it will not try to reduce it.
+fn defragment_page_fast(
+    page: &PageContent,
+    usable_space: u16,
+    freeblock_1st: usize,
+    freeblock_2nd: usize,
+) -> Result<()> {
+    turso_assert!(freeblock_1st != 0, "no free blocks");
+    if freeblock_2nd > 0 {
+        turso_assert!(freeblock_1st < freeblock_2nd, "1st freeblock is not before 2nd freeblock: freeblock_1st={freeblock_1st} freeblock_2nd={freeblock_2nd}");
+    }
+    const FREEBLOCK_SIZE_MIN: usize = 4;
+    turso_assert!(freeblock_1st <= usable_space as usize - FREEBLOCK_SIZE_MIN, "1st freeblock beyond usable space: freeblock_1st={freeblock_1st} usable_space={usable_space}");
+    turso_assert!(freeblock_2nd <= usable_space as usize - FREEBLOCK_SIZE_MIN, "2nd freeblock beyond usable space: freeblock_2nd={freeblock_2nd} usable_space={usable_space}");
+
+    let freeblock_1st_size = page.read_u16(freeblock_1st + 2) as usize;
+    let freeblock_2nd_size = if freeblock_2nd > 0 {
+        page.read_u16(freeblock_2nd + 2) as usize
+    } else {
+        0
+    };
+    let freeblocks_total_size = freeblock_1st_size + freeblock_2nd_size;
+
+    let cell_content_area = page.cell_content_area() as usize;
+
+    if freeblock_2nd > 0 {
+        // If there's 2 freeblocks, merge them into one first.
+        turso_assert!(freeblock_1st + freeblock_1st_size <= freeblock_2nd, "overlapping freeblocks: freeblock_1st={freeblock_1st} freeblock_1st_size={freeblock_1st_size} freeblock_2nd={freeblock_2nd}");
+        if freeblock_2nd + freeblock_2nd_size > usable_space as usize {
```

Comment:
What is the explanation on why we need 2 freeblocks and more importantly, `free_space < max_frag_bytes`?

Reply:
I get it now that I read the code, but... I wonder why 2 is the limit. I guess they had some timings.